src/backend/columnar/columnar_storage.c

```diff
@@ -0,0 +1,281 @@
+ * columnar_storage.c
+ *
+ * This file contains function definitions for reading and writing to SMgr to
+ * logical offsets (ignoring page headers).
+ *
+ * Copyright (c) 2016, Citus Data, Inc.
+ *
+ * $Id$
+ *
+ *-------------------------------------------------------------------------
+ */
+
+
+#include "postgres.h"
+
+#include "safe_lib.h"
+
+#include "access/heapam.h"
+#include "access/nbtree.h"
+#include "catalog/pg_am.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "storage/lmgr.h"
+#include "storage/smgr.h"
+#include "utils/guc.h"
+#include "utils/memutils.h"
+#include "utils/rel.h"
+#include "utils/relfilenodemap.h"
+
+#include "columnar/columnar.h"
+#include "columnar/columnar_storage.h"
+
+#define COLUMNAR_METAPAGE_BLOCKNO 0
+#define FIRST_LOGICAL_OFFSET COLUMNAR_BYTES_PER_PAGE
+#define LogicalOffsetIsValid(X) ((X) >= FIRST_LOGICAL_OFFSET)
+
+static void ReadFromBlock(Relation rel, BlockNumber blockno, uint32 offset,
+						  char *buf, uint32 len);
+static void WriteToBlock(Relation rel, BlockNumber blockno, uint32 offset,
+						 char *buf, uint32 len, bool clear);
+
+
+/*
+ * ColumnarMetapageInit - initialize a new metapage in an empty relation to
+ * the given contents.
+ */
+void
+ColumnarMetapageInit(Relation rel, ColumnarMetapage *metapage)
+{
+	LockRelationForExtension(rel, ExclusiveLock);
+
+#ifdef USE_ASSERT_CHECKING
+	RelationOpenSmgr(rel);
+	uint64 nblocks = smgrnblocks(rel->rd_smgr, MAIN_FORKNUM);
+	Assert(nblocks == 0);
+	Assert(sizeof(ColumnarMetapage) <= BLCKSZ - SizeOfPageHeaderData);
+#endif
+
+	/* create new page */
+	Buffer newBuffer = ReadBuffer(rel, P_NEW);
+	ReleaseBuffer(newBuffer);
+
+	WriteToBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				 (char *) metapage, sizeof(ColumnarMetapage), true);
+
+	UnlockRelationForExtension(rel, ExclusiveLock);
+}
+
+
+/*
+ * ColumnarMetapageRead - read the current contents of the metapage. No
+ * locks are held after return.
+ */
+void
+ColumnarMetapageRead(Relation rel, ColumnarMetapage *metapage)
+{
+	Assert(sizeof(ColumnarMetapage) <= BLCKSZ - SizeOfPageHeaderData);
+	ReadFromBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				  (char *)metapage, sizeof(ColumnarMetapage));
+}
+
+
+/*
+ * ColumnarMetapageReserveData - reserve logical data offsets for writing.
+ */
+uint64
+ColumnarMetapageReserveData(Relation rel, uint64 amount)
+{
+	LockRelationForExtension(rel, ExclusiveLock);
+
+	ColumnarMetapage metapage;
+	Assert(sizeof(ColumnarMetapage) <= BLCKSZ - SizeOfPageHeaderData);
+	ReadFromBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				  (char *) &metapage, sizeof(ColumnarMetapage));
+
+	/* make reservation starting on new page */
+	SmgrAddr prev_addr = logical_to_smgr(metapage.reserved_offset);
+	SmgrAddr initial = { 0 };
+	if (prev_addr.offset == SizeOfPageHeaderData)
+	{
+		initial.blockno = prev_addr.blockno;
+	}
+	else
+	{
+		initial.blockno = prev_addr.blockno + 1;
+	}
+	initial.offset = SizeOfPageHeaderData;
```

Comment:
maybe we can move this block into a function like `SmgrAddr GetReservationSmgrAddr(..)` to calculate `initial`